### PR TITLE
Use generics to simplify pkg/registry/flowcontrol/ensurer

### PR DESCRIPTION
**Pull request by MikeSpreitzer**
#### What type of PR is this?

/kind cleanup

#### What this PR does / why we need it:
This PR revises pkg/registry/flowcontrol/ensurer to use generics, which makes the code much more straightforward (as well as smaller).

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by MikeSpreitzer** [comment]
@MikeSpreitzer

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 1]
```diff
@@ -58,53 +59,34 @@ type EnsureStrategy interface {
...
-// The binding makes it unnecessary to have any type casts.
-// A bootstrap object is a mandatory or suggested config object,
-// with the spec that the code is built to provide.
-type BootstrapObjects interface {
-	typeName() string                         // the Kind of the objects
-	len() int                                 // number of objects
-	get(int) bootstrapObject                  // extract one object, origin 0
-	getExistingObjects() ([]deletable, error) // returns all the APF config objects that exist at the moment
-}
+type ObjectLocalOps[ObjectType configurationObject] interface {
```
Comment (thockin) [inline pkg/registry/flowcontrol/ensurer/strategy.go:65]: comments on all these would be nice?

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 2]
```diff
@@ -113,37 +95,70 @@ type configurationObject interface {
...
+type objectOps[ObjectType configurationObjectType] struct {
+	Client[ObjectType]
+	Cache[ObjectType]
+	deepCopy    func(ObjectType) ObjectType
+	replaceSpec func(ObjectType, ObjectType) ObjectType
+	specEqual   func(expected, actual ObjectType) bool
+}
+
+func NewObjectOps[ObjectType configurationObjectType](client Client[ObjectType], cache Cache[ObjectType],
+	DeepCopy func(ObjectType) ObjectType,
```
Comment (thockin) [inline pkg/registry/flowcontrol/ensurer/strategy.go:112]: function args are not usually `PublicStyle` but `privateStyle` names

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go [hunk 3]
```diff
@@ -210,23 +211,23 @@ func ensure(ctx context.Context, clientset flowcontrolclient.FlowcontrolV1beta3I
...
-	return ensurer.EnsureConfigurations(ctx, fsSuggesteds, ensurer.NewSuggestedEnsureStrategy())
+	fsOps := ensurer.NewFlowSchemaOps(clientset.FlowSchemas(), fsLister)
+	return ensurer.EnsureConfigurations(ctx, fsOps, flowcontrolbootstrap.SuggestedFlowSchemas, ensurer.NewSuggestedEnsureStrategy[*flowcontrolv1beta3.FlowSchema]())
 }
 
 func ensureMandatoryConfiguration(ctx context.Context, clientset flowcontrolclient.FlowcontrolV1beta3Interface, fsLister flowcontrollisters.FlowSchemaLister, plcLister flowcontrollisters.PriorityLevelConfigurationLister) error {
-	plcMandatories := ensurer.WrapBootstrapPriorityLevelConfigurations(clientset.PriorityLevelConfigurations(), plcLister, flowcontrolbootstrap.MandatoryPriorityLevelConfigurations)
-	if err := ensurer.EnsureConfigurations(ctx, plcMandatories, ensurer.NewMandatoryEnsureStrategy()); err != nil {
+	plcOps := ensurer.NewPriorityLevelConfigurationOps(clientset.PriorityLevelConfigurations(), plcLister)
+	if err := ensurer.EnsureConfigurations(ctx, plcOps, flowcontrolbootstrap.MandatoryPriorityLevelConfigurations, ensurer.NewMandatoryEnsureStrategy[*flowcontrolv1beta3.PriorityLevelConfiguration]()); err != nil {
```
Comment (thockin) [inline pkg/registry/flowcontrol/rest/storage_flowcontrol.go:225]: Do we want the pointer-ness in the type param, or is it safe assumed that internally?  E.g. `ensurer.NewMandatoryEnsureStrategy[*flowcontrolv1beta3.PriorityLevelConfiguration]()`  as the API and internally use `*` a lot?

**Review by thockin — COMMENTED** [review]
I am not super familiar with this code - @jpbetz do you know who might be?

**Review by thockin — COMMENTED** [review]
I am not super familiar with this code - @jpbetz do you know who might be?

**Comment by wojtek-t** [comment]
/assign

I'm familiar - I will take it :)

File: pkg/registry/flowcontrol/rest/storage_flowcontrol.go (hunk 3 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/rest/storage_flowcontrol.go:225]: The existing PR constrains the object type parameters to be subtypes of `runtime.Object` and `metav1.Object` --- and the pointer types meet those constraints, while the struct types do not.  I do not think there is an equally simple way to take the struct types as parameters while asserting that their pointers implement `runtime.Object` and `metav1.Object`.

**Comment by MikeSpreitzer** [comment]
The force-push to 1121b9ab450 responds to the review above and changes some stuff that was unnecessarily public to private.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 1 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go]: Yes, added them.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 2 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go]: Thanks, fixed.

**Comment by MikeSpreitzer** [comment]
The force-push to 3848ede350b is a rebase onto master.

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 4]
```diff
@@ -58,112 +59,138 @@ type EnsureStrategy interface {
...
+	replaceSpec func(ObjectType, ObjectType) ObjectType
+	specEqual   func(expected, actual ObjectType) bool
+}
+
+func NewObjectOps[ObjectType configurationObjectType](client client[ObjectType], cache cache[ObjectType],
+	deepCopy func(ObjectType) ObjectType,
+	replaceSpec func(ObjectType, ObjectType) ObjectType,
+	specEqual func(expected, actual ObjectType) bool,
+) ObjectOps[ObjectType] {
+	return objectOps[ObjectType]{client: client, cache: cache,
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/strategy.go:122]: nit: let's initialize one field per line:
```
return objectPos[ObjectType]{
  client: client,
  cache: cache,
  deepCopy: deepCopy,
  ...
}
```

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 5]
```diff
@@ -58,112 +59,138 @@ type EnsureStrategy interface {
...
+func (s *strategy[ObjectType]) Name() string {
 	return s.name
 }
 
-func (s *strategy) ShouldUpdate(wah wantAndHave) (updatable, bool, error) {
-	current := wah.getHave()
-
-	if current == nil {
-		return nil, false, nil
+func (s *strategy[ObjectType]) ShouldUpdate(objectOps objectLocalOps[ObjectType], current, bootstrap ObjectType) (ObjectType, bool, error) {
```
Comment (wojtek-t) [inline pkg/registry/flowcontrol/ensurer/strategy.go:168]: Given it returns updated object, I suggest renaming this function.

`UpdateIfNeeded` maybe?

**Review by wojtek-t — COMMENTED** [review]
Just some small nits - other than that LGTM.

**Comment by MikeSpreitzer** [comment]
The force-push to 92a35f5bca9 is the renaming and linebreak suggested in review above.

File: pkg/registry/flowcontrol/ensurer/strategy.go [hunk 6]
```diff
@@ -58,112 +59,138 @@ type EnsureStrategy interface {
 	// ok: true if auto update is required, otherwise false
 	// err: err is set when the function runs into an error and can not
 	//      determine if auto update is needed.
-	ShouldUpdate(wantAndHave) (revised updatable, ok bool, err error)
+	ShouldUpdate(objectOps objectLocalOps[ObjectType], current, bootstrap ObjectType) (revised ObjectType, ok bool, err error)
```
[the remark this answers is not in the document]
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go]: Yes, that is a better name (regardless of how the revision is conveyed).  Changed (in all relevant places).

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 4 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go]: Done.

File: pkg/registry/flowcontrol/ensurer/strategy.go (hunk 5 above)
Comment (MikeSpreitzer) [reply pkg/registry/flowcontrol/ensurer/strategy.go]: Done, along with all the other copies of that name.

**Pull request merged**